net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,395 @@
+
+static inline u32 calc_alloc_len(const struct gkprioq_sched_data *q)
+{
+	u32 total_alloc_len = 0;
+	int prio;
+	
+	for (prio = 0; prio < MAX_PRIORITY + 1; prio++) 
+		total_alloc_len += (q->prio_array[prio])->limit;
+
+	return total_alloc_len;
+}
+
+static inline u16 calc_new_high_prio(const struct gkprioq_sched_data *q)
+{
+	int prio;
+
+	if (q->highest_prio > (HALF_MAX_PRIORITY + 1)) {
+		for (prio = q->highest_prio - 1; prio > HALF_MAX_PRIORITY + 1; 
+		                                 prio--) {
+			if ((q->prio_array[prio])->q.qlen > 0)
+				return prio;
+		}
+	}
+
+	if ((q->prio_array[0])->q.qlen > 0) /*Treating as best effort queue*/
+		return 0;
+
+	for (prio = HALF_MAX_PRIORITY + 1; prio > 0; prio--) {
+		if ((q->prio_array[prio])->q.qlen > 0)
+			return prio;
+	}
+
+	return 0; /*GK queue is empty, return default highest priority setting*/
+
+}
+
+static int gkprioq_enqueue(struct sk_buff *skb, struct Qdisc *sch,
+			   struct sk_buff **to_free)
+{
+	struct Qdisc *your_queue;
+	int prio, ret, err, wlen, gk_len;
+	u16 your_prio; 
+
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	wlen = skb_network_offset(skb);
+
+	switch(tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		wlen += sizeof(struct iphdr);
+		if (!pskb_may_pull(skb, wlen) || 
+		    skb_try_make_writable(skb, wlen))
+			goto drop;
+
+		skb->tc_index = ipv4_get_dsfield(ip_hdr(skb))
+			& ~INET_ECN_MASK;
+		break;
+
+	case htons(ETH_P_IPV6):
+		wlen += sizeof(struct ipv6hdr);
+		if (!pskb_may_pull(skb, wlen) || 
+		    skb_try_make_writable(skb, wlen))
+			goto drop;
+
+		skb->tc_index = ipv6_get_dsfield(ipv6_hdr(skb))
+			& ~INET_ECN_MASK;
+		break;
+
+	default: 
+		skb->tc_index = 0;
+		break;
+	}
+
+	your_prio = skb->tc_index;  /*DSCP field set as priority*/
+
+	if (your_prio > q->highest_prio) /*Checking if a new highest priority*/	 	q->highest_prio = your_prio;
+			
+	your_queue = q->prio_array[your_prio];
+
+	gk_len = calc_alloc_len(q);
+	if (gk_len < q->max_limit) {
+		err = fifo_set_limit(your_queue, 
+				    your_queue->limit + q->init_limit);
+		if (err) 
+			return err;
+	
+		q->init_limit += 1; /* Fast forward resource allocation*/
+	
+		ret = qdisc_enqueue(skb, your_queue, to_free);
+		if (ret != NET_XMIT_SUCCESS) {
+			if (net_xmit_drop_count(ret))
+				qdisc_qstats_drop(sch);
+		return ret;
+		}
+
+		qdisc_qstats_backlog_inc(sch, skb);
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+
+	} else if (your_prio == (0 || 1) && your_queue->limit > 0) {
+	
+		qdisc_drop(skb, sch, to_free);	
+	
+		/* Priority 0 queue is the default case queue and is not allowed 		 * to borrow or be borrowed from. It is treated as a best effort 		 * queue. Priority 1 queue is the least priority and so can't		         * borrow from anyone.  
+		 */
+	
+		qdisc_qstats_backlog_inc(sch, skb);
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+
+	} else {
+
+	/* Queues need to borrow resources from other queues */	
+		
+	if (your_queue->q.qlen == your_queue->limit && your_queue->limit > 0) {
```

Comment:
Can you describe the case that this large if statement is controlling? We don't have anything analogous to this in the Gatekeeper code.